src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -360,6 +362,10 @@ void BaseCouplingScheme::secondExchange()
         _isTimeWindowComplete = true;
         _timeWindowStartTime += _computedTimeWindowPart;
         _computedTimeWindowPart = 0.0; // reset window
+        _totalDriftCorrection += _driftCorrection;
+        if (abs(_totalDriftCorrection) > math::NUMERICAL_ZERO_DIFFERENCE) {
+          PRECICE_WARN("preCICE has corrected the time by a total of {}. This is often necessary, if you are using very many substeps per time window over multiple time windows.", _totalDriftCorrection);
```

Comment:
I do not know if it makes sense add: preCICE has corrected the time by a total of {}. This leads to an additional time integration error which might effect the accuracy of the solution. 

For me the connection between this error and moving the time step is far from trivial. This might also lead to difficult to debug errors when the time integration error is small.

Reply:
> For me the connection between this extra error and moving the time step is far from trivial. This might also lead to difficult to debug errors when the time integration error is small.

Absolutely. What we are doing here will probably have side effects, but our current approach is even more problematic (see #1866). I think we will learn more about the situation in the future and ideally this will lead to a fix that generally works without errors or the necessity to raise warnings.

The ideal solution from my perspective would be some additional synchronization between the preCICE time and the solver time. However, this requires considerable API changes.

Reply:
> This leads to an additional time integration error which might effect the accuracy of the solution.

A solution could be:

> In case you are computing the length of the final timestep of the window, try to directly use the return value of `getMaxTimeStepSize()` instead.

Reply:
> In case you are computing the length of the final timestep of the window, try to directly use the return value of `getMaxTimeStepSize()` instead.

That's a very good idea. I directly implemented it. We now don't have to implement any additional functionality with unclear outcome. Instead we raise an error, if something looks wrong and the user has to fix it. Fixing it is relatively easy and the user also knows what's going on. See https://github.com/BenjaminRodenberg/oscillator-example/commit/6741dcf5edf7c7c51532ab67fc94df47e720b925.

Reply:
My intent was to append it to the warning, but if the total drift is a valid enough criterium to bail out, then even better.

I would avoid referring to a GitHub issue though. Let's use a full text instead. Given that it is an error now, message length doesn't matter :wink:

Reply:
Well... this is something we should actually explain in the documentation with a code example. Maybe a bit too much for an error message. But also touching the website... Not sure that I'm currently in the right mood to open another box :laughing: